WFindOnWebMenu: Menu for to find track properties in online music databases

fatihemreyildiz opened the pull request:
This is a draft PR.

As we mentioned on the Zulip #gsoc stream Web search Box feature topic, I wanted to add a small feature, that can help users to save time. A shortcut to find Artists on different services. For now I tried on SoundCloud.

I am facing with an issue. I couldn't fetch artist name on wtrackmenu. I have tried few things but they didn't work. At last, i tried something and it returns an error, which is signal and slot is not equal. I have looked at it, couldn't find a solution. 

For the future of this feature, there can be one main row called as "Find on..." Then many rows could be there. Such as, "Find the artist on Discogs", "Find the track on LastFm" etc. In my estimation to make it bigger feature there should be something like search related track menu.

ronso0 commented on src/widget/wtrackmenu.h:
```diff
@@ -10,6 +10,7 @@
 #include "library/trackprocessing.h"
 #include "preferences/usersettings.h"
 #include "track/beats.h"
+#include "track/track.h"
```
instead of including the entire header here, can Track be forward-declared?

fatihemreyildiz replied:
Yes. I have deleted this line.

ronso0 commented:
> I am facing with an issue. I couldn't fetch artist name on wtrackmenu. I have tried few things but they didn't work. At last, i tried something and it returns an error, which is signal and slot is not equal. I have looked at it, couldn't find a solution.

> which is signal and slot is not equal

Did you read https://doc.qt.io/qt-5/signalsandslots.html or similiar ressources on signals/slots?
[QAction::triggered](https://doc.qt.io/qt-5/qaction.html#triggered)
Look into the slot above `WTrackMenu::slotOpenInFileBrowser` for how to collect the track.
If you enable the feature also for the tracks table WTracktableView you need to decide wether you want the feature to work for multiple tracks (multple web calls) or just for a single selected track. Like, the action could be disabled in `WTrackMenu::updateMenus` if multiple tracks are selected.
Actually, in general I suggest to take a thorough look at similiar WTrackMenu actions.

Also, please rebase this onto main to resolve the merge conflicts.

daschuer commented on src/widget/wtrackmenu.cpp:
```diff
@@ -949,6 +963,12 @@ void WTrackMenu::slotOpenInFileBrowser() {
 }
 
+void WTrackMenu::slotFindOnSoundcloud(const Track& track) {
+    const auto artistName = track.getArtist();
```
I think you can use     
`TrackPointer pTrack = getFirstTrackPointer();`
to obtain the track.

fatihemreyildiz replied:
That worked.

Thank you :)

ronso0 commented on src/widget/wtrackmenu.cpp:
```diff
@@ -963,10 +967,10 @@ void WTrackMenu::slotOpenInFileBrowser() {
-    QDesktopServices::openUrl(QUrl("https://soundcloud.com/search/people?q="));
+void WTrackMenu::slotFindOnSoundcloud() {
+    const TrackPointer pTrack = getFirstTrackPointer();
+    QString artistName = pTrack->getArtist();
```
This could be empty. Use album artist as fallback?
If that is also empty, ask the user? Or just open the soundcloud search and let the user type there?

fatihemreyildiz replied:
Yeah, album artist and asking the user might be good option. I can implement that, If it's okay.

Right now it opens SoundCloud with a blank page.

daschuer replied:
I think we need to % encode the artist name. 

Opening SoundCloud with a blank page sounds good enough for me, to realize that the artist is empty. 

What exactly do we want to ask the user? What will be the alternative? We may gray out the menu entry or open the track property dialog to add an artist.

daschuer replied:
> I think we need to % encode the artist name.

Oh I just realized that QUrl is smart enough to do it automatically.

ronso0 replied:
> What exactly do we want to ask the user? What will be the alternative? We may gray out the menu entry or open the track property dialog to add an artist.

Graying out the menu action is not helpful IMO. Initially I thought of a simple textbox dialog asking for the track artist, apply that to the track, then attempting to open soundcloud again. But that's too cumbersome UX-wise and to program for this first step.

Since an empty soundcloud query would not solve the missing tag issue we could show a dialog
> This track has no artist or album artist tag.
Try again after filling the tag(s) in the track properties dialog or inline in the tracks table.
[ Open track properties] [ Okay ]

uklotzde replied:
Keep it simple without requiring any additional user input, see the "Search related Tracks" feature. If I need to type I could open a browser myself.

Mixxx should not become a launcher for external applications. Opening the file explorer is a notable but useful exception. If a feature can't be integrated properly into Mixxx it should not become part of Mixxx. The feature as currently planned is already an edge case IMHO.

fatihemreyildiz replied:
I have looked at "Search Related Tracks" feature it is so nice. When there is no artist info. It doesn't show the find artist row. This feature can be like this too, so no user input needed and also can be extended, such as find the artist, albumartist, track on lastfm discogs or musicbrainz in a single menu such as "Search on web".

Opening the file browser is a must and it is very useful. Launching the web browser can be useful too. I think for some users It can be helpful to find the artists on different services. If they are looking for information on the web a lot about the artist, such as latest releases, albums etc. It can save their time.

ronso0 commented:
> Also, please rebase this onto main to resolve the merge conflicts

...and make CI work.

fatihemreyildiz commented:
> > I am facing with an issue. I couldn't fetch artist name on wtrackmenu. I have tried few things but they didn't work. At last, i tried something and it returns an error, which is signal and slot is not equal. I have looked at it, couldn't find a solution.
> 
> > which is signal and slot is not equal
> 
> Did you read https://doc.qt.io/qt-5/signalsandslots.html or similiar ressources on signals/slots? [QAction::triggered](https://doc.qt.io/qt-5/qaction.html#triggered) Look into the slot `WTrackMenu::slotOpenInFileBrowser` for how to collect the track. If you enable the feature also for the tracks table WTracktableView you need to decide wether you want the feature to work for multiple tracks (multple web calls) or just for a single selected track. Like, the action could be disabled in `WTrackMenu::updateMenus` if multiple tracks are selected. Actually, in general I suggest to take a thorough look at similiar WTrackMenu actions.
> 
> Also, please rebase this onto main to resolve the merge conflicts.

Thanks for the feedback.

I have read the exact page. I couldn't resolve the problem. I was facing with this error when I tried to get Track info by reference. Otherwise, there was not any error. But this time I couldn't get the trackinfo.

I tried to collect the track by looking to other slots but couldn't collect it in that way too.

Now I got the artist by pointer.

As you mentioned, I have disabled the multiple tracks and rebase the branc onto main.

ronso0 commented:
> I have read the exact page. I couldn't resolve the problem.

Signal and slot need matching arguments (or you use a slot that doesn't expect arguments at all, like your current slot).
QAction::triggered only has `bool triggered`, though your previous slot implementation expected `Track` as argument.

fatihemreyildiz commented:
> > I have read the exact page. I couldn't resolve the problem.
> 
> Signal and slot need matching arguments (or you use a slot that doesn't expect arguments at all, like your current slot). QAction::triggered only has `bool triggered`, though your previous slot implementation expected `Track` as argument.

Ah, That's right. Now I got it why I was having this error and couldn't resolve it. 

Thank you :)

uklotzde commented on src/widget/wtrackmenu.cpp:
```diff
@@ -949,6 +967,12 @@ void WTrackMenu::slotOpenInFileBrowser() {
+void WTrackMenu::slotFindOnSoundcloud() {
+    const TrackPointer pTrack = getFirstTrackPointer();
+    QString artistName = pTrack->getArtist();
+    QDesktopServices::openUrl(QUrl("https://soundcloud.com/search/people?q=" + artistName));
```
I recommend not to use naive string composition for creating URLs, see `QUrl::setQuery()`.

Please also externalize all the business logic into separate classes/files, don't bother `WTrackMenu` with this extra stuff. It is already too big.

fatihemreyildiz replied:
Thanks for the feedback, I will take a look at QUrl::setQuery() and work on business logic.

fatihemreyildiz replied:
Hey again,

I was trying to move all the logic into other classes and I was working on `QUrl::setQuery()`. I just have a question, 

I have a function, which takes 2 arguments one is the URL itself and the second one is query (in this case, Artist, Album artist, Album, Track title). My Url is `https://www.last.fm/search/artists?q=` and the artist is `Eminem`. 

When I call the method `QUrl::setQuery()`, it works just fine, the Url is created. When the Url is opened I realized the opened Url is `https://www.last.fm/search/artists?Eminem` which opens a page without any results. 'q=' part is gone. 

Should I pass all the queries with the prefix "=q" into my function or should I use another method which fits into key value pattern?

Because when I looked into documentation it says about 'setQuery'.
> This function is useful if you need to pass a query string that does not fit into the key-value pattern

Swiftb0y replied:
From the Qt doc, For me it seems like you should use [QUrlQuery](https://doc.qt.io/qt-5/qurlquery.html) and then use the `QUrl::setQuery(QUrlQuery)` overload. QUrlQuery:
> The QUrlQuery class provides a way to manipulate a key-value pairs in a URL's query. [More...](https://doc.qt.io/qt-5/qurlquery.html#details)

The QUrl class does much more than just concatenate string. When you pass it an a URL with a query and then call `setQuery`, it will overwrite the previous query, which is why `setQuery("Eminem")` results in the previous query (`q=` being overwritten). Since you are dealing with key-value pairs here, use `QUrlQuery`.

fatihemreyildiz replied:
Thanks for your feedback. 

Now I realized that I should use both the QUrlQuery and setQuery. For two different things.

If pagination is simple and I can directly open the artist-track-album page. I can simply use setQuery. Because there is no key and value pairs. There will be no queries. 

If we can't open artist-track-album page directly and if we need to do search on the service, I should use QUrlQuery, to have key q= and the value artist-track-album name.

fatihemreyildiz committed:
wtrackmenu: Searchbox feature

daschuer commented:
You branch looks like that in gitk. 
![grafik](https://user-images.githubusercontent.com/1777442/170842577-5148b597-ffce-424a-b95e-f6ebf52630c5.png)
It was made on top of a previous feature branch, merged to 2.3. Since this one will go to main it is that difficult to resolve. 

Luckily git has a solution: 
* First stash you uncommited work `git stash`
* That fetch the latest main branch `git fetch upstream`
* Rebase the commits belonging to this branch on top of main `git rebase --onto upstream/main HEAD~3`
* (`HEAD~3` is the third commit below HEAD 214dbbcb1a5fd9494bed9dca1479c822fb876347, the old base, that will not be moved to main) 
* Reslove conflicts, by editing the regions with ">>>>>>>>>>>>>"
* Stage the edits `git add src/widget/wtrackwidgetgroup.cpp` the same with the others
* `git rebase --continue`
* Resolve conflicts again until rebase is finish   
* Restore you working files `git stash pop`

I use for conflicts SmartGit which has a nice three files diff view for that.

daschuer commented:
Ah you have solved the problem and squashed everything to a single commit. Great.

daschuer commented:
Works good!

fatihemreyildiz commented:
Hey everyone!

We have discussed with my mentor about this PR and feature. Either it is merged or not it can be a playground to learn the Mixxx code base for me, so I would like to work on it a bit more.

One of the most important thing is WTrackMenu class is too big and we don't want to bother it a lot. To do that we want to use factory design pattern for FindOn feature. So with that, we can have separate files. Expand this feature with different websites. It can be turned on or off on the preferences menu. 

But I am wondering how users can use this feature, open the default browser? Shouldn't they right click on the track, on the TrackMenu and then choose either one of the websites? I should still use the WTrackMenu class to call FindOn right? I don't want to bother WTrackMenu class a lot. I wonder how will it look like on the user perspective. 

Instead of QUrl I will try to use QUrl::setQuery().

Should I also keep going on this PR, or on a brand new branch and PR?

daschuer commented:
IMHO As long as this state does not break something in main we could merge in steps of less perfect solutions. But wat means good enough here? Ideas?

daschuer commented:
I am unsure where the FindOn feature should be hooked. My original request, a Discogs lookup for possible genres and the first time release date would be also a good fit in the track property dialog. The same is also true for cover art look up ... Both locations will finally work. Ideas?

ronso0 commented:
> I am unsure where the FindOn feature should be hooked

Finding the artist/track site on audio platforms (Soundcloud, Spotify) is IMO about gathering additional info on the fly or looking for other tracks by that artist (not necessary similiar/matching tracks).
So to make that accessible from the track table / deck it should go to the track menu, though that is indeed already pretty crowded. FWIW I'm repeatedly annoyed by the top-level AutoDJ action, I think we should consider making the track menu configurable in the preferences (child of **Library** page)

Looking for covers and genre should IMO happen in the Track Properties.

fatihemreyildiz committed:
CMakeLists: WFindOn added

fatihemreyildiz committed:
WFindOnMenu: Separate Find On Menu class added.

fatihemreyildiz committed:
WTrackMenu: Qaction deleted, Find On class added

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+        Discogs
+    };
+
+    void openTheBrowser(const QString& serviceUrl,
```
openInBrowser?

The serviceUrl should not be part of the public API.

daschuer replied:
I think this function can be moved to the anonymous namespace in the cpp file, because it does not use any members from "this"

fatihemreyildiz replied:
I tried to make the URL private. But since I am using this method in WTrackMenu, this is a bit difficult. Or is it too easy and am I missing something?

I made private the Url in an another way. I defined a function in anonymous namespace that can return URL's according to chosen service, I am calling this function in openInBrowser, to return the urls from anonymous namespace. This time openInBrowser gets 2 arguments, which is the service and the query. But this time I needed to make "enum Service" public. So I am not sure which should be public and the other one private, is it okay to make Enum Service public?

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+    void openTheBrowser(const QString& serviceUrl,
+            const QString& query);
+
+    void createAllServices(const Track& track);
```
"create all services"? I don't understand what this method is supposed to do.

daschuer replied:
I agree, without a comment, it looks like this function created the whole service itself. 
How about createSubmenusForAllWebLookups() or such.  


By the way, in header we use /// comments that are picked up by doxygen and asigned to the following function which is able to automatically create an API decryption along with a class hierarchy.

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+    void createAllServices(const Track& track);
+
+  signals:
+    void triggerBrowser(
```
What's the difference between those 2 arguments with similar names and identical types? Very error prone.

daschuer replied:
maybe webLookupQuerry and porperty, or such. 
You may also consider to use enums to release the menu from implementation details.

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+            QWidget* parent = nullptr);
+    ~WFindOnMenu() override = default;
+
+    enum class Services {
```
Each value of this enum denotes a single service, not multiple. The type name is confusing.

uklotzde replied:
As long as this enum is not needed by consumers of the menu make it private or hide it in the .cpp file in an anonymous namespace.

uklotzde commented:
The term "find on" is unintuitive for a feature that actually performs a "web search".

uklotzde commented:
Please ensure that your code passes the CI checks, even if it is only a _Draft PR_.

daschuer commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,218 @@
+}
+
+inline QString searchUrlDiscogsGen() {
+    return QString("https://www.discogs.com/search/?");
```
My Discogs use case is searching for: 
`https://www.discogs.com/search/?q=The%20Cure%20-%20Friday%20I%27m%20In%20Love&type=release`
This gives good Result in this case. 
`https://www.discogs.com/search/?type=release&title=Dancing+Queen&artist=ABBA` 
Finds only exact matches, which is probably not what we want, but OK in this case.  

I also like the "Artist" only search, to look up track from a particular artist. 

I am not sure about the "Album" only and the "Track" only search. This reveals similar named albums and tracks. 
Not too useful if the album has a common name like "Greatest Hits"   

Do we have a real use case for this?

Do we have other interesting use cases?

fatihemreyildiz replied:
We can have a specific search for discogs, which can have multiple queries. So that will give us better results.

daschuer replied:
I think the use case question is valid for all lookups:
* For Title only: find remixes 
* For Album only: no idea, remove? 
* Artist + Track: lookup additional info.

fatihemreyildiz replied:
Should Title only search configured to find remixes? Or should it stay still?

I think, that would be nice to have just album to look up if there are any other tracks in that album. But I can remove it too.

Artist + Track lookup would be so nice, we can have exact search and not lost in similar artists and tracks. But I have checked now, only discogs have search with multiple queries. Others accepts just a query which is "q=", artist and track search is done over different pagination.

daschuer replied:
> Should Title only search configured to find remixes? Or should it stay still?

Is this possible? Currently it reveals all titles with the same name, the original Track is at a random place in the result. 

> I think, that would be nice to have just album to look up if there are any other tracks in that album. 

For that use case you have to look up the Artist + Album. Else you the desired album might be listed among many by-catch.

> Artist + Track lookup would be so nice, we can have exact search and not lost in similar artists and tracks. 

I think when we just append Artist + Title, the result will Be good for most services. Check it out.

daschuer replied:
Sounds good. 

The Remix addition works only sometimes. I think If we just using the title, remixes will be in the result.

Especially on SoundCloud it works perfect.

fatihemreyildiz commented:
Thanks for the feed backs, I will resolve the issues asap. I will be more careful with the naming, not gonna use inline functions and can do better search with discogs.

fatihemreyildiz committed:
wtrackmenu wfindonmenu: CI Fix clazy / pre-commit

fatihemreyildiz committed:
WFindOnMenu: Inline Functions are deleted

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+        Discogs
+    };
+
+    void openTheBrowser(const QString& serviceUrl,
```
The menu is controlled by signals and should not expose implementation details as public member functions.

fatihemreyildiz committed:
WFindOnMenu: Enum naming changed moved private

fatihemreyildiz committed:
WFindOnMenu: composeActionText add, order changed

fatihemreyildiz commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,218 @@
+        const Track& track,
+        QString serviceTitleDisplay,
+        Services serviceTitle) {
+    serviceMenu = new QMenu(this);
```
Since WFindOnMenu is owned by the Q_OBJECT? So something like 
`pserviceMenu = new WFindOnMenu(this);`
would solve this issue?

daschuer replied:
auto pServiceMenu = make_parented<QMenu>(this);

Since WFindOnMenu is inherited from QObject, it will delete all child QObjects once it is deleted, during the inherited QObject destructor.

parented_ptr makes sure that this is the case and will be still the case after refactoring.

daschuer commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+            const QString& query);
+
+  private:
+    void createService(QMenu* serviceMenu,
```
serviceMenu is already a "service related submenu" the name should reveal this. 
This function is also a candidate to be moved into the anonymous namespace. 

Maybe the function can be called "populateWebLookUpQueries" ... It depends in which direction you will turn the interface.

fatihemreyildiz committed:
WFindOnMenu: namings refactored for all instances.

fatihemreyildiz commented:
I have tried be careful with the naming, Tried to fix the CI issue I have compiled before the push. Changed few basic things such as, moved enum to private, deleted inline functions, changed the order of the action buttons.

Now I will try to make few changes in order to make it better.
I will move Url to private and move some methods to anon namespace.
Add way better web search, such as remixes.
Better Discogs queries.
I'll use parented pointer.
At last, change the name of the feature such as Web Search, Search on Web, Open in web browser etc. For this a survey on Zulip would be nice IMHO.

daschuer commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,45 @@
+            QWidget* parent = nullptr);
+    ~WFindOnMenu() override = default;
+
+    void openInBrowser(const QString& serviceUrl,
```
Here you should add 
`public slots:`

fatihemreyildiz replied:
Thank you. I have added it in my last commit.

daschuer commented:
Thank you. This has become mature for a merge.
We just need to consider the final name and rename everything accordingly.

Doy you have in IDE that has support for it?
Eclipse does, but is not 100% working.

fatihemreyildiz committed:
WFindOnMenu: Url hidden. Search with artist added.

fatihemreyildiz commented:
> Thank you. This has become mature for a merge.
We just need to consider the final name and rename everything accordingly.
Doy you have in IDE that has support for it?
Eclipse does, but is not 100% working.

Thank you for your review. For the Final name, should we have a survey on Zulip?

I am using Visual Studio Code, I think I can do it quickly. But first we need a name I guess. I can also try Eclipse too.

fatihemreyildiz committed:
WFindOnMenu: CI Fix clazy

uklotzde commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,279 @@
+    return trackAlbumOrTitle + QStringLiteral(" ") + artistTitle;
+}
+
+QString composePrefixAction(WFindOnMenu::TrackProperty trackProperty) {
```
Are these supposed to be i18n strings that need to be translated? Then this won't work.

fatihemreyildiz replied:
I think this needs to be translated. In order to make them i18n strings, According to QT Doc: [Writing Source Code for Translation](https://doc.qt.io/qt-5/i18n-source-translation.html) I should return QString instead of QStringLiteral, right?

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,64 @@
+    enum class TrackProperty {
+        Artist,
+        ArtistTitle,
+        ArtistAlbum,
```
This terminology is confusing, because there actually is a property that is named "album artist". If you intend to mean `ArtistAndAlbum` then please be explicit about it.

If the enum variants denote a conjunction of one or more individual properties then `TrackSearchProperties` would be a better name.

fatihemreyildiz replied:
For ArtistTitle and ArtistAlbum, I wasn't sure to use "and". Next time I will be more explicit and precise. I will make them "ArtistAndAlbum", "ArtistAndTitle" 

Right now I am fully aware about the enum naming. If they denote single property, they should be named Singular and if they denote conjunction, they should be named Plural. Thanks for this review! I will make it "TrackSearchProperties" as you suggested.

uklotzde commented on src/widget/wfindonmenu.h:
```diff
@@ -0,0 +1,64 @@
+            WFindOnMenu::TrackProperty trackProperty,
+            const QString& query);
+
+    void createAllSubmenusForWebLookups(const Track& track);
```
Why does this need to be a public slot? It is invoked as an ordinary member function.

uklotzde replied:
"find on", "web lookups", now we have already 2 different terms for the same thing. The name of the member functions should also not redundantly repeat the purpose of the class.

Imho a name like `populateFromTrackProperties()` would be more suitable here. That this task involves sub menus is an implementation detail that should not be revealed and doesn't matter for the caller.

fatihemreyildiz replied:
> Why does this need to be a public slot? It is invoked as an ordinary member function.

I was calling these functions to connect signals in WTrackMenu.

> The name of the member functions should also not redundantly repeat the purpose of the class.

I will be also careful about that too.  

> Imho a name like populateFromTrackProperties() would be more suitable here. That this task involves sub menus is an implementation detail that should not be revealed and doesn't matter for the caller.

This function is generating submenus named with the "services-websites" and it is called in WTrackMenu. IMHO, I would like to use "addMenusForWebSearch". Since it is called in WTrackMenu, that would be easier to understand what this function does.

The naming also can be createAllMenusForWebSites or createMenuForServices. I think we should decide it later on what this feature will be called.

BTW this function "createAllSubmenusForWebLookups" is calling another functions right now, they are named "populateWebLookUpQueries", this can be named as "populateFromTrackProperties" which is more like they are intended to do. 

What do you think?

fatihemreyildiz replied:
I realized that I am repeating the purpose of the class by using "addMenusForWebSearch". "WebSearch" part is revealing the purpose.

Since I named the websites as Service. This function creates QMenu named with service names. Should I make it something like "addSubmenusForServices"?

uklotzde commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,279 @@
+        TrackProperty trackProperty,
+        const QString& query) {
+    const QString serviceUrl = getServiceUrl(service, trackProperty);
+    QUrlQuery qurlquery;
```
Repeating the type name in lowercase is not a good choice for naming a variable. Even the `Q` prefix is repeated.

fatihemreyildiz replied:
Oh, Okay. I will be more careful about not naming by repeating the type name. Thanks for this review. I will named it as "query".

uklotzde commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,279 @@
+        const Track& track,
+        const QString& serviceTitle,
+        Service service) {
+    const auto artistTitle = track.getArtist();
```
What's an "artist title"??

fatihemreyildiz replied:
I will change it as "primaryArtist".

daschuer replied:
Why not just artist?

Sometimes I have artists like
"Foo feat. Bar"

I don't think we will be able to extract the primary artist.

fatihemreyildiz replied:
Just artist is okay :+1: .

I thought for later, we might want to add something with "album artist" as a second artist lookup. So if this become wanted addition, we can think about adding this primaryArtist and secondaryArtist as album artist.

daschuer replied:
That route is not easy because the album artist ist often not even populate. It is interesting If it is a featuring Album like Apocalyptica, recording their track with different vocal artists.

fatihemreyildiz replied:
> It is interesting If it is a featuring Album like Apocalyptica, recording their track with different vocal artists.

Would be nice to have then, but should I implement it right now or the later?

uklotzde commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,279 @@
+                    trackTitle,
+                    pServiceMenu,
+                    WFindOnMenu::TrackProperty::Title);
+            addActionsTrackTitle(service,
```
addActionsTrackTitle() that is invoked with a parameter ArtistTitle? Sorry, you lost me.

fatihemreyildiz replied:
Sorry. Same as before, all ArtistTitle's are going to be "primaryArtist".

uklotzde commented:
> Thank you. This has become mature for a merge.

I disagree.

fatihemreyildiz commented:
Thank you for the reviews @uklotzde. I will try to take care of them. Before another Push, I will also change the feature name to avoid misunderstandings.

daschuer commented on src/widget/wfindonmenu.cpp:
```diff
@@ -0,0 +1,279 @@
+
+QString composePrefixAction(WFindOnMenu::TrackProperty trackProperty) {
+    if (trackProperty == WFindOnMenu::TrackProperty::Title) {
+        return QStringLiteral("Title");
```
```suggestion
        return QObject::tr("Title");
```
Should do the trick.

fatihemreyildiz replied:
Thank you! :)

fatihemreyildiz committed:
Feature renamed. Naming refactored.

fatihemreyildiz committed:
wfindonwebmenu: artistwithtitle deleted.

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,282 @@
+    } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Artist) {
+        return QObject::tr("Artist");
+    } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle) {
+        return QObject::tr("Title && Artist");
```
The escaped & can be a translator trap.
How about  +?

fatihemreyildiz replied:
Oh okay, didn't notice that & would be an issue to translating. I will change it as + .

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,282 @@
+
+QString searchUrlSoundCloudAlbum = QStringLiteral("https://soundcloud.com/search/albums?");
+
+QString searchUrlDiscogsGen = QStringLiteral("https://www.discogs.com/search/?");
```
We can refine Discogs search, by utilize the detailed search.
But it is fine to postpone it to a follow up PR

fatihemreyildiz replied:
Exactly, I was waiting for reviews before i dive into detailed Discogs search. Because there are many queries needed for Discogs, which is different than other services.

daschuer reviewed (approved):
Thank you for the renaming. LGTM I will do some manual tests tonight.
@uklotzde should have the final word.

daschuer commented:
Is this still in a draft state?

fatihemreyildiz commented:
I think not anymore. Because beginning of this PR was for to get help to get Track Property in WTrackMenu. Now we have different web search menus and actions in separate classes and files.

daschuer commented:
Ok, than remove the draft state.

uklotzde reviewed (changes requested):
@daschuer Please review more thoroughly before proposing to merge.

fatihemreyildiz committed:
const url add. redundants deleted. func refactored

daschuer commented:
This now looks like that: 
![grafik](https://user-images.githubusercontent.com/1777442/174503043-063358c5-bfe5-4e1e-9833-d74645dbe0e4.png)

I suggest to put the "Title + Artist" or maybe "Artist" on top. I consider both as the most common use case.

fatihemreyildiz committed:
WFindOnWebMenu: pre-commit fix. Action reordered.

fatihemreyildiz commented:
All the requested changes are done :+1:

fatihemreyildiz commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,282 @@
+        const QString& serviceTitle,
+        Service service) {
+    const auto artist = track.getArtist();
+    pServiceMenu = new QMenu(this);
```
Thanks for this feedback. I have missed that somehow but I find out where I was mistaken. It helped me a lot!

uklotzde commented:
@daschuer I just noticed that this adds another top-level entry to the track menu. An item that opens just an exterbal web browser and leaves Mixxx doesn't belong there. It should be hidden in an submenu, e.g. in _Metadata >_.

daschuer commented:
A Submenu of metadata will also work for me.

fatihemreyildiz committed:
WTrackMenu: FindOnWeb moved as submenu of Metadata

fatihemreyildiz committed:
WTrackMenu: pre-commit fix

uklotzde commented:
> If it seems okay, I will try to make this feature with Factory Method separate the services and so later on another services can simple added.

Not sure what you are referring to.

What I noticed is that the code contains lots of switch statements with code that only differs in values. This is a strong indicator that the design doesn't fit the problem at hand. Code written in this style is hardly maintainable or extendable. It should be refactored by applying OO or whatever principles.

@daschuer I am not planning to provide any guidance here. As long as the code in question is separated and doesn't interfere with core functionality I do not care enough.

daschuer commented:
OK, thank you anyway. We like to get this version merged, since it is already useful. Do you agree? Is anything missing?

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+    }
+}
+
+const QString searchUrlSoundCloudArtist = QStringLiteral("https://soundcloud.com/search/people?");
```
Naming conventions are still violated -> `kSearchUrlSoundCloudArtist`

fatihemreyildiz replied:
Done :+1:. I have changed all constant URL to start with k.

What is k stands for? I am just asking to learn. I have checked the coding guidelines and google for naming conventions of constants couldn't find it.

fatihemreyildiz replied:
Makes sense. Thank you

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+    switch (service) {
+    case WFindOnWebMenu::Service::Discogs:
+        return searchUrlDiscogsGen;
+        break;
```
redundant break

fatihemreyildiz replied:
Deleted.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+        const QString& serviceTitle,
+        Service service) {
+    const auto artist = track.getArtist();
+    m_pServiceMenu = new QMenu(this);
```
This could be memory leak if `m_pServiceMenu` is re-initialized while not being owned by someone else.

Why a member variable? I don't see that it is used anywhere?

fatihemreyildiz committed:
wfindonmenu: naming changed, parentedptr added

daschuer commented:
Please merge from upstream/main
There is a trivial conflict because of the recent addition of the analyze menu.

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+    }
+}
+
+const QString ksearchUrlSoundCloudArtist = QStringLiteral("https://soundcloud.com/search/people?");
```
```suggestion
const QString kSearchUrlSoundCloudArtist = QStringLiteral("https://soundcloud.com/search/people?");
```
(Sorry for keeping nitpicking ...)

fatihemreyildiz replied:
It's totally okay. I have must missed that.

fatihemreyildiz committed:
wfindonwebmenu: constant naming fix

fatihemreyildiz committed:
Merge remote-tracking branch 'upstream/main' into websearchbox

daschuer commented on src/widget/wfindonwebmenu.h:
```diff
@@ -0,0 +1,61 @@
+            QMenu* pServiceMenu,
+            TrackSearchProperties trackSearchProperty);
+
+    QMenu* pServiceMenu;
```
This one is unused now.

fatihemreyildiz committed:
wfindonwebmenu: pServiceMenu deleted.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+        QMenu* pServiceMenu,
+        TrackSearchProperties trackSearchProperty) {
+    const QString prefixAction = composePrefixAction(trackSearchProperty);
+    switch (service) {
```
This switch/case seems to be completely redundant and could be removed.

uklotzde commented on src/widget/wtrackmenu.cpp:
```diff
@@ -185,6 +186,34 @@ void WTrackMenu::createMenus() {
+                    if (pTrack) {
+                        m_pFindOnMenu->addSubmenusForServices(*pTrack);
+                    }
+                    m_pFindOnMenu->setEnabled(
```
If pTrack is null the menu contents become stale and the menu remains enabled.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+        QMenu* pServiceMenu,
+        TrackSearchProperties trackSearchProperty) {
+    const QString prefixAction = composePrefixAction(trackSearchProperty);
+    switch (service) {
```
This switch/case seems to be completely redundant and could be removed.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+        QMenu* pServiceMenu,
+        TrackSearchProperties trackSearchProperty) {
+    const QString prefixAction = composePrefixAction(trackSearchProperty);
+    switch (service) {
```
This switch/case seems to be completely redundant and could be removed.

fatihemreyildiz committed:
wfindonwebmenu: redundant switch&functions deleted

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,269 @@
+    urlQuery.addQueryItem("q", query);
+    QUrl url(serviceUrl);
+    url.setQuery(urlQuery);
+    QDesktopServices::openUrl(url);
```
since the failure is unexpected, using Verify_or_Drbugassert() should be used.

uklotzde replied:
However. But at least a warning log message should be logged to allow tracing unexpected failures.

fatihemreyildiz replied:
What would be the best use-case in this scenario? is qWarning okay, or should qDebug be used? Should URL be in the log warning?

daschuer replied:
Oh I see the debug assertion is a already above. I did some test and I was not able to fail openUrl() even with obscure urls. 

So I guess an error only happens when the users system is miss-configured.
https://github.com/qt/qtbase/blob/42e13b7c61693ca95e119106d4f6dbd2bcf2308d/src/gui/util/qdesktopservices.cpp#L195 
And many of those cases have already a qWarning when digging deeper. 

Conclusion: A normal qWarning() including the URL is sufficient. 
Maybe just technical: 
"QDesktopServices::openUrl() failed for <URL>"

fatihemreyildiz replied:
I have tried to check URL's in two different ways.
First with [QUrl::isValid()](https://doc.qt.io/qt-6/qurl.html#isValid) which returns always false because the propoerties have white spaces and they are not encoded, that is because it is always returning false.
Second is (Which I used for now) [QUrl::isEmpty()](https://doc.qt.io/qt-6/qurl.html#isEmpty) which returns always true because the URL is not empty at all. But IDK if this one makes wanted validation.

My question is should I keep using isEmpty() with existing configuration, can it make validation or should I make encoding with white spaces and use isValid()? 

---

While I was writing my question, I have tried it again, I realised that isValid() is returning true with the existing URL's which everything was fine in the first place. Then I realised that VERIFY_OR_DEBUG_ASSERT() is executed if the condition is false, so I should of use it like VODA(isValid()) instead of VODA(!isValid())

So long story short, I guess, It's better to use isValid() with the technical qWarning().

uklotzde replied:
All URLs produced by this code must be valid. Anything else would be a programming error, hence a DEBUG_ASSERT without a log is appropriate.

But QDesktopServices::openUrl() might fail for other reasons out of our control when launching the application, even if the URL is valid.

fatihemreyildiz replied:
I think right now with the last commits URL is validated. 

And for the QDesktopServices::openUrl() we can not do anything, right?

uklotzde replied:
Please log a warning message inlcuding the URL as already suggested, wrapped into VERIFY_AND_DEBUG_ASSERT.

fatihemreyildiz committed:
wfindonwebmenu: url-check & switch added

fatihemreyildiz committed:
wtrackmenu: empty pTrack submenu not show added.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,167 @@
+    }
+    {
+        const auto trackTitle = track.getTitle();
+        const auto artistWithTrackTitle = composeSearchQuery(trackTitle, artist);
```
It is confusing to compute a value that might get discarded by the following if statement. Abort early.

fatihemreyildiz replied:
Solved :+1:

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,167 @@
+    }
+    {
+        const auto album = track.getAlbum();
+        const auto artistWithAlbum = composeSearchQuery(album, artist);
```
Once again.

fatihemreyildiz replied:
Solved :+1:

uklotzde reviewed (changes requested):
@fatihemreyildiz Please only request a re-review after resolving all comments.

fatihemreyildiz commented:
> @fatihemreyildiz Please only request a re-review after resolving all comments.

I thought I solved all, sorry for misunderstanding. Which comment was not solved? URL validation?

fatihemreyildiz committed:
wfindonwebmenu: if statement value aborted early

fatihemreyildiz committed:
wfindonwebmenu: url validation refactored.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,163 @@
+    urlQuery.addQueryItem("q", query);
+    QUrl url(serviceUrl);
+    url.setQuery(urlQuery);
+    VERIFY_OR_DEBUG_ASSERT(url.isValid()) {
```
The warning message is wrong at this point. QDesktopServices::openUrl() wasn't event invoked.

fatihemreyildiz replied:
openUrl() has a URL validation inside. [openUrl() URL validation](https://github.com/qt/qtbase/blob/42e13b7c61693ca95e119106d4f6dbd2bcf2308d/src/gui/util/qdesktopservices.cpp#L210)

So my check was duplication of an existing check and redundant. So instead, it is better to check if "QDesktopServices::openUrl()" fails itself and warn the developers with "DEBUG_ASSERT(false)", so it logged before Mixxx quits the assertion.

Also about the first check, I was checking if the URL is empty, this wasn't the best validation for Url's and also it was redundant too.  

I will now commit the URL validation again, I hope this time we will have a well working URL validation.

fatihemreyildiz committed:
wfindonwebmenu: url validation added.

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,162 @@
+                pServiceMenu,
+                WFindOnWebMenu::TrackSearchProperties::Artist);
+    }
+    if (!trackTitle.isEmpty()) {
```
This condition is not sufficient for both actions

fatihemreyildiz committed:
wfindonmenu: if condition added for actions

uklotzde commented on src/widget/wtrackmenu.cpp:
```diff
@@ -185,6 +186,38 @@ void WTrackMenu::createMenus() {
+                            (pTrack->getArtist().isEmpty() &&
+                                    pTrack->getAlbum().isEmpty() &&
+                                    pTrack->getTitle().isEmpty());
+                    if (pTrack && !checkTrackProperties) {
```
If `pTrack = nullptr` then Mixxx has already crashed at this point.

uklotzde commented:
> OK I just had another look to the code and I think all issues are solved.

Nope.

uklotzde commented on src/widget/wtrackmenu.cpp:
```diff
@@ -185,6 +186,38 @@ void WTrackMenu::createMenus() {
+                [this] {
+                    m_pFindOnMenu->clear();
+                    const auto pTrack = getFirstTrackPointer();
+                    const bool checkTrackProperties =
```
The code is not self-explaining and needs to be improved. During the review I am not speaking for myself but for everyone who will read this code in the future.

-> `static bool WFindOnWebMenu::hasEntriesForTrack(const Track& track);`

fatihemreyildiz committed:
wfindonwebmenu:  bool hasEntriesForTrack added

daschuer commented on src/widget/wtrackmenu.cpp:
```diff
@@ -897,6 +893,11 @@ void WTrackMenu::updateMenus() {
+
+    if (featureIsEnabled(Feature::FindOnWeb)) {
+        const auto pTrack = getFirstTrackPointer();
+        m_pFindOnMenu->setEnabled(!WFindOnWebMenu::hasEntriesForTrack(*pTrack));
```
In C++ it is assumed that a reference can't be null. But this is only the case if the programmer takes care of it. 
In this case the pTrack can be null and the assumption is void. So please ass a check for null to fix that.

fatihemreyildiz replied:
So as far as I understood, before we set enable if the track has entries, we should check the pointer is null or not.  I think we should do that twice because we set enable two times first while creating the menu second while updating the menu right?

daschuer replied:
Yes right. But I suggest to give the pointer to WFindOnWebMenu::hasEntriesForTrack and return false in that case. 
This way we have not much code duplication.

uklotzde replied:
@daschuer Passing a shared pointer is conceptually wrong if you do not plan to store it.

uklotzde replied:
No. It should be a const-ref that is also used by all other public methods. Passing a pointer is pointless and wrong. The whole menu would be obsolete without a valid `Track` object. Please don't confuse this!

daschuer replied:
It is hard to guess which solution do you have in mind without more details. I don't understand why passing a raw pointer is pointless and wrong and doing the null check inside hasEntriesForTrack()

My suggestion was this: 

```suggestion
        m_pFindOnMenu->setEnabled(!WFindOnWebMenu::hasEntriesForTrack(pTrack.get()));
```

With a null check inside hasEntriesForTrack() 

Do you wish this solution? 

```suggestion
        const bool hasEntries = pTrack ? WFindOnWebMenu::hasEntriesForTrack(*pTrack) : false; 
        m_pFindOnMenu->setEnabled(!hasEntries);
```

or do you have something else in mind?

uklotzde replied:
WFindOnWebMenu must not be bothered with nullptr checks. This would be ridiculous.

uklotzde replied:
It does not need a pointer as an argument nor does it store one.

daschuer replied:
Since we are talking here about a static function this argument is void. But anyhow, this is a matter of taste finally. You have not yet commented my alternatives. Is the second one OK to you?

uklotzde replied:
This is not a matter of taste and it doesn't matter if this is a static function or not. It's about scopes and responsibilities. The 2nd alternative is correct, because it keeps the nullptr check where it belongs. This also matches my initial proposal.

uklotzde replied:
`const auto enableMenu = pTrack && WFindOnWebMenu::hasEntriesForTrack(*pTrack);`

The name `hasEntries` is inappropriate in this context.

uklotzde replied:
You will hopefully notice that the implementation of `hasEntriesForTrack()` is also still wrong.

daschuer replied:
What do you suggest? hasActions / hasSubmenus ...

daschuer replied:
Ah, scratch that, you mean enableMenu instead of hasEntries for the local variable. Ok.

uklotzde replied:
> Yes, the whole PR suffers from unconditinal use of auto. This makes the code hard to review.

I disagree strongly. This isn't even an issue.

daschuer replied:
You speak in riddles. Please give us more hints why the function is still wrong. 

> I disagree strongly. 

That is you personal preference and not part of our style guide. It even lists the following cases as bad:  
```const auto artist = track.getArtist();```
This is because in a review you cannot be certain what type auto is.

uklotzde commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,178 @@
+    const auto artist = track.getArtist();
+    const auto trackTitle = track.getTitle();
+    const auto album = track.getAlbum();
+    return (artist.isEmpty() && album.isEmpty() && trackTitle.isEmpty());
```
@daschuer If all those properties are empty the menu _has entries_. Sure?

Instead of complaining about the usage of `auto` I recommend to focus on what really matters.

The expression `track.getArtist().isEmpty()` also doesn't reveal the type of the intermediate result and would be permitted according to our rules. Your "the type is out of sight" arguments are invalid.

daschuer replied:
Good catch.

fatihemreyildiz committed:
FindOnWeb: enableMenu added.

fatihemreyildiz committed:
wfindonwebmenu: type change

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,177 @@
+        const QString& query) {
+    const QString serviceUrl = getServiceUrl(service, trackSearchProperty);
+    QUrlQuery urlQuery;
+    urlQuery.addQueryItem("q", query);
```
query is not the whole query, it is only the queryValue, or trackPropertyValue or ...

daschuer commented on src/widget/wfindonwebmenu.h:
```diff
@@ -0,0 +1,53 @@
+    static bool hasEntriesForTrack(const Track& track);
+
+    void openInBrowser(WFindOnWebMenu::Service service,
+            WFindOnWebMenu::TrackSearchProperties trackSearchProperty,
```
trackSearchProperties, because it can also be two

daschuer reviewed (changes requested):
I have found an improvement regarding the connections and some naming nit-picks.

fatihemreyildiz committed:
wfindonwebmenu: naming changed

fatihemreyildiz committed:
wfindonwebmenu: Redundant connect deleted

daschuer commented on src/widget/wfindonwebmenu.h:
```diff
@@ -0,0 +1,48 @@
+  private:
+    void openInBrowser(WFindOnWebMenu::Service service,
+            WFindOnWebMenu::TrackSearchProperties trackSearchProperties,
+            const QString& query);
```
```suggestion
            const QString& queryValue);
```

daschuer commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,177 @@
+            composeActionText(prefixAction, queryValue),
+            this,
+            [this, service, trackSearchProperties, queryValue]() {
+                this->openInBrowser(service, trackSearchProperties, queryValue);
```
```suggestion
                openInBrowser(service, trackSearchProperties, queryValue);
```

fatihemreyildiz committed:
wfindonwebmenu: requested changes applied

Swiftb0y commented on src/widget/wfindonwebmenu.cpp:
```diff
@@ -0,0 +1,177 @@
+void WFindOnWebMenu::addSubmenusForServices(const Track& track) {
+    WFindOnWebMenu::populateFromTrackProperties(
+            track,
+            QString(tr("SoundCloud")),
+            Service::SoundCloud);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("LastFm")), Service::LastFm);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("Discogs")), Service::Discogs);
```
I don't think these have to / should be translatable.

fatihemreyildiz replied:
Thank you for the feedback. 

I will change them to QStringLiteral.

Swiftb0y commented on src/widget/wtrackmenu.cpp:
```diff
@@ -185,6 +186,25 @@ void WTrackMenu::createMenus() {
+        connect(m_pFindOnMenu,
+                &QMenu::aboutToShow,
+                this,
+                [this] {
```
this will cause a warning on C++20. You'll want to capture this by copy. 
```suggestion
                [*this] {
```

uklotzde replied:
Are your sure? QWidget are not copyable and  you always want to pass the pointer. We use this pattern everywhere else, but never `[*this]`.

fatihemreyildiz replied:
What should I do? I'm a bit confused. [this] or [*this]? 

I used [this] because I saw it in other lambdas in codebase.

daschuer replied:
With [this] the this pointer and it is captured by value, so everything is OK. 
This way all member variables are captured by reference via the this pointer as usual in member functions.
 
[*this] will capture the whole object by value. For my understanding changes to the member variables are not visible in other member variables. This is surprising, and and the copy an unnecessary overhead.

fatihemreyildiz replied:
Thanks for the explanation. Every feedback is golden for me.

This time I will go with `[this]`

Swiftb0y replied:
I thought copies of QObjects are just reference-counted internally? But sure we can capture by reference if we can guarantee that `this` won't become dangling.

Swiftb0y reviewed (commented):
Can't really comment on this code otherwise as I have 0 experience with QWidgets or the mixxx library in general.

fatihemreyildiz committed:
wfindonwebmenu: non translatable names changed.

Swiftb0y commented:
@daschuer is volunteering to address (or mentor @fatihemreyildiz to fix) any issues introduced by this PR. LGTM from me so I'll go ahead and merge.

uklotzde commented:
The intermediate commits should have been stashed before merging. Now the history contains lots of unneeded diffs and broken commits.

Swiftb0y commented:
I'm sorry my mistake. I suppose squashing and force-pushing to `main` is not an option now?

uklotzde commented:
> I'm sorry my mistake. I suppose squashing and force-pushing to `main` is not an option now?

I don't think so. Mistakes happen, don't worry.
